### PR TITLE
FISH-6603

### PR DESCRIPTION
**Pull request by luiseufrasio**
## Description
New 0-day vulnerability exploit using ROOT context root deployments

### New tests
New case in StandardContextValveTest.normalizeURLTest

### Testing Performed
Steps in https://payara.atlassian.net/browse/FISH-6603
using Payara\appserver\extras\payara-micro\payara-micro-distribution\target\payara-micro.jar

### Testing Environment
Zulu JDK 1.8_212 on Ubuntu 18.04 with Maven 3.8.4

### CODE REVIEW

**Comment by luiseufrasio** [comment]
jenkins test please

**Comment by breakponchito** [comment]
Jenkins test please

**Review by breakponchito — APPROVED** [review]
LGTM

File: appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java [hunk 1]
```diff
@@ -286,13 +287,19 @@ protected String normalize(String path) {
...
-        if(rv.indexOf("./") == 0) {
+        if (rv.indexOf("./") == 0) {
             rv = rv.replaceFirst("./", "/");
         }
+        // has WEB-INF or META-INF
+        if (rv.toUpperCase().indexOf("WEB-INF") > 0) {
+            rv = "/" + rv.substring(rv.toUpperCase().indexOf("WEB-INF"));
+        } else if (rv.toUpperCase().indexOf("META-INF") > 0) {
+            rv = "/" + rv.substring(rv.toUpperCase().indexOf("META-INF"));
+        }
```
Comment (pdudits) [inline appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java:302]: How hot is this path, what percentage of requests would hit it? If almost every one, then two uppercase conversions is a little bit wasteful and case-insentivive regexp would be more efficient, though less readable.

**Review by pdudits — APPROVED** [review]

File: appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java (hunk 1 above)
Comment (luiseufrasio) [reply appserver/web/web-core/src/main/java/org/apache/catalina/core/StandardContextValve.java]: Performance improved and now checking if it contains "/WEB-INF/" or ends with "/WEB-INF".
The same check for "META-INF

**Comment by luiseufrasio** [comment]
jenkins test please

**Comment by breakponchito** [comment]
Jenkins test please

**Comment by breakponchito** [comment]
Jenkins test please

**Review by breakponchito — APPROVED** [review]
LGTM

**Pull request merged**